operator-common/src/main/java/io/strimzi/operator/common/AbstractOperator.java

```diff
@@ -488,31 +488,38 @@ public AtomicInteger getResourceCounter() {
      * @param reconciliation reconciliation to use to update the resource state metric
      * @param ready if reconcile was successful and the resource is ready
      */
-    private void updateResourceState(Reconciliation reconciliation, boolean ready) {
-        Tags metricTags = Tags.of(
+    private void updateResourceState(Reconciliation reconciliation, boolean ready, Throwable cause) {
+        String key = reconciliation.namespace() + ":" + reconciliation.kind() + "/" + reconciliation.name();
+
+        Tags metricTagsWithReason = Tags.of(
                 Tag.of("kind", reconciliation.kind()),
                 Tag.of("name", reconciliation.name()),
-                Tag.of("resource-namespace", reconciliation.namespace()));
+                Tag.of("resource-namespace", reconciliation.namespace()),
+                Tag.of("failure-reason", cause == null ? "none" : cause.getMessage()));
 
         T cr = resourceOperator.get(reconciliation.namespace(), reconciliation.name());
+
+        Optional<Meter> gauge = metrics.meterRegistry().getMeters()
+                .stream()
+                .filter(meter -> meter.getId().getName().equals(METRICS_PREFIX + "resource.state") &&
+                        meter.getId().getTags().contains(Tag.of("kind", reconciliation.kind())) &&
+                        meter.getId().getTags().contains(Tag.of("name", reconciliation.name())) &&
+                        meter.getId().getTags().contains(Tag.of("resource-namespace", reconciliation.namespace()))
+                ).findFirst();
+
+        if (gauge.isPresent()) {
+            // remove metric so it can be re-added with new tags
+            metrics.meterRegistry().remove(gauge.get().getId());
+            resourcesStateCounter.remove(key);
+            log.debug("{}: Removed metric " + METRICS_PREFIX + "resource.state{}", reconciliation, key);
+        }
```

Comment:
why this part is not in the "else" condition when the cr is null?

Reply:
Basically we do remove the metric and only if the cr is not null, re-create it.